node-neigh: Fix node removal and invalid neigh entry due to buggy arping response correlation

brb opened the pull request:
Please see commit messages.

For backporters: please ignore the last commit.

Fixes #14693.

brb committed:
node-neigh: Do not inc neighbor refcount for the same node

(*linuxNodeHandler).NodeUpdate() can be called multiple times for the
same remote node. This could lead to the neighbor refcount being
incremented if the nextHop of the node had not changed. The consequence
of the invalid refcount is that upon (*linuxNodeHandler).NodeDelete(),
a related neighbor entry might not be deleted.

Fix this by checking whether the node has been previously considered for
arpinging. An alternative solution would be to pass the firstAddition
param to insertNeighbor(). However, as we are going to merge
refreshNeighbor(), the taken approach is more robust and requires less
assertions in the upcoming merged method.

Fixes: 6c06c51926b ("node: Remove permanent ARP entry when remote node is deleted")
Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
node-neigh: Merge {insert,refresh}Neighbor()

This commit fixes two issues via refactoring.

Previously, refreshNeighbor() was very similar to insertNeighor().  This
commit refactors insertNeighbor() so that it can be invoked by
refreshNeighbor(), and thus DRY.

Besides refactoring, this commit fixes the following issues:

- The linuxNodeHandler.mutex is being held during the whole
  insertNeighor() execution. Previously, refreshNeighbor() took the lock
  twice which opened a window for a race when
  n.neighByNextHop[nextHopStr] could have been set to an obsolete neigh
  if insertNeighor() and refreshNeighbor() were executed at the same
  time. Also, the old arping library sends requests via a raw
  socket, so it might mix responses resulting in mixed neigh entries
  (will be fixed and explained in a separate PR).
- refreshNeighbor() won't increment the refcount. We expect
  insertNeighor() to do so. Otherwise, we risk with both methods
  incrementing.

Fixes: 5ec4d51f980 ("daemon, node: refresh neighbor by sending arping periodically")
Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
node-neigh: Do not refresh neighbors if node is not ready

Wait until the node is initialized. When it's not, insertNeighbor() is
not called, thus there is no neighbor entries to be refreshed.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
node-neigh: Fix insertNeighbor logging

- Use "node-neigh" subsystem field to better identify sometimes cryptic
  arping related warnings.
- Set proper IP addr for nexthop.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
arping: Improve response correlation

The arping library sends ARP requests via a raw socket (proto set to
ARP). It means that the socket can can receive any ARP packet.

Previously, the library was filtering responses based on whether a
packet is ARP response AND response source IP addr == arpinged IP addr.

Improve the filter by adding a check whether destination IP addr == ARP
request source IP addr.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -609,11 +609,45 @@ func (n *linuxNodeHandler) insertNeighbor(newNode *nodeTypes.Node, ifaceName str
+						logfields.LinkIndex:    neigh.LinkIndex,
+					}).WithError(err).Warn("Failed to remove neighbor entry")
+				}
+				delete(n.neighByNextHop, nextHopStr)
```
Should we still delete it even if the `netlink` returned an error?

brb replied:
Yeah, as otherwise nobody would remove it from the map.

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -588,7 +588,7 @@ func (n *linuxNodeHandler) getSrcAndNextHopIPv4(nodeIPv4 net.IP, ifaceName strin
 
 // Must be called with linuxNodeHandler.mutex held.
-func (n *linuxNodeHandler) insertNeighbor(newNode *nodeTypes.Node, ifaceName string) {
+func (n *linuxNodeHandler) insertNeighbor(ctx context.Context, newNode *nodeTypes.Node, ifaceName string, refresh bool) {
```
Please add documentation to the parameters of this function, it's not clear to me anymore what are certain parameters used for.

aanm reviewed (commented):
Some small nits, doesn't require my 2nd review.

rbtcollins commented on pkg/datapath/linux/arp/ping.go:
```diff
@@ -98,7 +98,8 @@ func (p *pinger) resolve(ip net.IP) (net.HardwareAddr, error) {
 		}
 
-		if resp.Operation != layers.ARPReply || !bytes.Equal(resp.SourceProtAddress, ip.To4()) {
+		if resp.Operation != layers.ARPReply || !bytes.Equal(resp.SourceProtAddress, ip.To4()) ||
```
Applying deMorgan here would make this a little easier to read:

```
if resp.Operation == layers.ARPReply && bytes.Equal(resp.SourceProtAddress, ip.To4()) &&
bytes.Equal(resp.DstProtAddress, p.ip.To4()) {
```

pchaigno commented on pkg/datapath/linux/node_linux_test.go:
```diff
@@ -1025,6 +1025,11 @@ func (s *linuxPrivilegedIPv4OnlyTestSuite) TestArpPingHandling(c *check.C) {
 	err = linuxNodeHandler.NodeAdd(nodev1)
 	c.Assert(err, check.IsNil)
+	// Insert the same node second time. This should not increment refcount for
+	// the same nextHop. We test it when check that NodeDelete has removed the
```
Nit: missing word in second sentence.

pchaigno commented on pkg/datapath/linux/node.go:
```diff
@@ -562,11 +562,11 @@ func (n *linuxNodeHandler) getSrcAndNextHopIPv4(nodeIPv4 net.IP, ifaceName strin
 
 	if len(routes) == 0 {
-		return nil, nil, fmt.Errorf("remote node IP is non-routable")
+		return nil, nil, fmt.Errorf("Remote node IP is not routable. Connectivity to pods on that node may be unavailable.")
```
I think this commit is reverting all changes from the two previous commits by mistake.

brb replied:
:facepalm:

pchaigno replied:
I don't think this is resolved. Most of commit `node-neigh: Unconditionally remove neigh entry from BPF map` is just reverts of the two previous commits.

brb replied:
:facepalm: x2 (having two workdirs doesn't help)

brb committed:
node-neigh: Unconditionally remove neigh entry from BPF map

Otherwise, we risk to silently blackhole packets instead of
emitting DROP_NO_FIB when inserting a neigh via netlink fails.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
